h5py/h5t.pyx

```diff
@@ -1106,7 +1106,17 @@ cdef class TypeCompoundID(TypeCompositeID):
         else:
             if sys.version[0] == '3':
                 field_names = [x.decode('utf8') for x in field_names]
-            typeobj = dtype({'names': field_names, 'formats': field_types, 'offsets': field_offsets})
+            if len(field_names) > 0:
+                collated_fields = zip(field_names, field_types, field_offsets)
+                ordered_fields = sorted(
+                    collated_fields, key=operator.itemgetter(2))
+                field_names, field_types, field_offsets = \
+                    [list(t) for t in zip(*ordered_fields)]
```

Comment:
Almost, it would even work in Python I think, but not in Cython. Since `field_names`, `field_types`, and `field_offsets` are declared to be `list`, not `tuple`, this assignment will fail at runtime unless we explicitly convert the result of `zip(*ordered_fields)` to `list`. Alternatively we could have made them tuples instead of lists. I didn't end up doing that to avoid invasively changing code elsewhere.